### PR TITLE
Backup Bug Fixes

### LINKED ISSUE #4502: [Backup] Supplying non-existent name or natively formatted name for container causes stack trace
Trying to run:
`az backup protection backup-now -g tjp-ig -v vault1 --retain-until 01-01-2020 -c container1 -i item1`

Yields:
```
'list' object has no attribute 'name'
Traceback (most recent call last):
  File "/opt/az/lib/python3.6/site-packages/azure/cli/main.py", line 36, in main
    cmd_result = APPLICATION.execute(args)
  File "/opt/az/lib/python3.6/site-packages/azure/cli/core/application.py", line 212, in execute
    result = expanded_arg.func(params)
  File "/opt/az/lib/python3.6/site-packages/azure/cli/core/commands/__init__.py", line 377, in __call__
    return self.handler(*args, **kwargs)
  File "/opt/az/lib/python3.6/site-packages/azure/cli/core/commands/__init__.py", line 626, in _execute_command
    reraise(*sys.exc_info())
  File "/opt/az/lib/python3.6/site-packages/six.py", line 693, in reraise
    raise value
  File "/opt/az/lib/python3.6/site-packages/azure/cli/core/commands/__init__.py", line 603, in _execute_command
    result = op(client, **kwargs) if client else op(**kwargs)
  File "/opt/az/lib/python3.6/site-packages/azure/cli/command_modules/backup/custom.py", line 198, in backup_now
    container_type, item_type)
  File "/opt/az/lib/python3.6/site-packages/azure/cli/command_modules/backup/custom.py", line 145, in show_item
    items = list_items(client, resource_group_name, vault_name, container_name, container_type, item_type)
  File "/opt/az/lib/python3.6/site-packages/azure/cli/command_modules/backup/custom.py", line 159, in list_items
    return [item for item in paged_items if item.properties.container_name.lower() in container.name.lower()]
  File "/opt/az/lib/python3.6/site-packages/azure/cli/command_modules/backup/custom.py", line 159, in <listcomp>
    return [item for item in paged_items if item.properties.container_name.lower() in container.name.lower()]
AttributeError: 'list' object has no attribute 'name'
```

In this case, it was not clear to me that a container and item had been created by the `enable-for-vm` command, so I assumed that these names were the names that would be created. When I realized it expected an existing name, I tried the container name obtained by running the `backup container list` command. However this yields the same stack trace. This AttributeError should be caught and handled gracefully.  Additionally, it is perfectly fine to use a shorthand notation (like the VM name in this case) but:

1. It should be clearly documented in help
2. The parameter should definitely accept the format of name that the container object actually uses.

---

### Environment summary

**Install Method:** How did you install the CLI? (e.g. pip, interactive script, apt-get, Docker, MSI, nightly)  
Answer here: WGET

**CLI Version:** What version of the CLI and modules are installed? (Use `az --version`)  
Answer here: azure-cli (2.0.17+dev)

acr (2.0.11+dev)
acs (2.0.15+dev)
appservice (0.1.16+dev)
backup (1.0.0+dev)
batch (3.1.3+dev)
billing (0.1.4+dev)
cdn (0.0.8+dev)
cloud (2.0.7+dev)
cognitiveservices (0.1.7+dev)
command-modules-nspkg (2.0.1+dev)
component (2.0.7+dev)
configure (2.0.10+dev)
consumption (0.1.4+dev)
container (0.1.10+dev)
core (2.0.16+dev)
cosmosdb (0.1.12+dev)
dla (0.0.11+dev)
dls (0.0.13+dev)
eventgrid (0.1.3+dev)
extension (0.0.2+dev)
feedback (2.0.6+dev)
find (0.2.6+dev)
interactive (0.3.9+dev)
iot (0.1.11+dev)
keyvault (2.0.10+dev)
lab (0.0.10+dev)
monitor (0.0.9+dev)
network (2.0.14+dev)
nspkg (3.0.1+dev)
profile (2.0.12+dev)
rdbms (0.0.6+dev)
redis (0.2.8+dev)
resource (2.0.14+dev)
role (2.0.11+dev)
servicefabric (0.0.3+dev)
sql (2.0.11+dev)
storage (2.0.15+dev)
vm (2.0.14+dev)

**OS Version:** What OS and version are you using?  
Answer here: WSL

**Shell Type:** What shell are you using? (e.g. bash, cmd.exe, Bash on Windows)  
Answer here: Bash on Windows

### LINKED ISSUE #4960: Azure Backup Recoverypoint Listing Error
Trying to run te following command:

`az backup recoverypoint list -v $vaultName -g $resourceGroup -c $itemName -i $itemName`
Getting the Following Error:
Might be Connected to the Backed up Item doesnt exist anymore
It Works on some of the Items but not on All.

> az : ERROR: 'list' object has no attribute 'id'
> In Zeile:27 Zeichen:33
> + ... ointsJson = az backup recoverypoint list -v "********" -g "******* ...
> +                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>     + CategoryInfo          : NotSpecified: (ERROR: 'list' o... attribute 'id':String) [], RemoteException
>     + FullyQualifiedErrorId : NativeCommandError
>  
> Traceback (most recent call last):
>   File "C:\Program Files (x86)\Microsoft SDKs\Azure\CLI2\lib\site-packages\azure\cli\main.py", line 36, in main
>     cmd_result = APPLICATION.execute(args)
>   File "C:\Program Files (x86)\Microsoft SDKs\Azure\CLI2\lib\site-packages\azure\cli\core\application.py", line 216, in 
> execute
>     result = expanded_arg.func(params)
>   File "C:\Program Files (x86)\Microsoft SDKs\Azure\CLI2\lib\site-packages\azure\cli\core\commands\__init__.py", line 377, 
> in __call__
>     return self.handler(*args, **kwargs)
>   File "C:\Program Files (x86)\Microsoft SDKs\Azure\CLI2\lib\site-packages\azure\cli\core\commands\__init__.py", line 620, 
> in _execute_command
>     reraise(*sys.exc_info())
>   File "C:\Program Files (x86)\Microsoft SDKs\Azure\CLI2\lib\site-packages\six.py", line 693, in reraise
>     raise value
>   File "C:\Program Files (x86)\Microsoft SDKs\Azure\CLI2\lib\site-packages\azure\cli\core\commands\__init__.py", line 602, 
> in _execute_command
>     result = op(client, **kwargs) if client else op(**kwargs)
>   File "C:\Program Files (x86)\Microsoft SDKs\Azure\CLI2\lib\site-packages\azure\cli\command_modules\backup\custom.py", line 
> 230, in list_recovery_points
>     container_uri = _get_protection_container_uri_from_id(item.id)
> AttributeError: 'list' object has no attribute 'id'

---

### Environment summary

Install Method (e.g. pip, interactive script, apt-get, Docker, MSI, edge build) / CLI version (`az --version`) / OS version / Shell Type (e.g. bash, cmd.exe, Bash on Windows)  
MSI
PowerShell
Windows 1709
azure-cli (2.0.21)
```
```

### PR DESCRIPTION
**Pull request by dragonfly91**
Backup Bug Fixes
---
This PR fixes #4960. 

Also partially fixes #4502. Supplying non-existent name for container will NOT throw a stack trace exception now. However, supplying a natively formatted name will still fail. Will address this in a future release.

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [x] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

### Command Guidelines

- [ ] Each command and parameter has a meaningful description.
- [ ] Each new command has a test.

(see [Authoring Command Modules](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules))

### CODE REVIEW

**Comment by tjprescott** [comment]
Due to recent changes to the CLI core, you will need to rebase your changes against the latest in the dev branch.

File: src/command_modules/azure-cli-backup/HISTORY.rst [hunk 1]
```diff
@@ -3,6 +3,11 @@
 Release History
 ===============
 
+1.0.5
++++++
+* Bugfix: Supplying non-existent name or natively formatted name for container causes stack trace.
```
Comment (tjprescott) [inline src/command_modules/azure-cli-backup/HISTORY.rst]: This looks like the issue name, what was the fix and to which command(s)?  The see the fix for if a non-existent name is supplied... does it actually support using the natively formatted name?

File: src/command_modules/azure-cli-backup/HISTORY.rst [hunk 2]
```diff
@@ -3,6 +3,11 @@
...
+* Enhancement: Now 'Health Status' of an item is visible in its table view.
```
Comment (tjprescott) [inline src/command_modules/azure-cli-backup/HISTORY.rst]: This should not which command(s) now display health status.  For example:

```
* `backup container update`: "Health Status" is now visible in the default table view.
```

File: src/command_modules/azure-cli-backup/azure/cli/command_modules/backup/custom.py [hunk 3]
```diff
@@ -610,6 +621,23 @@ def _get_resource_name_and_rg(resource_group_name, name_or_id):
...
+def _validate_container(container):
+    _validate_object(container, "Container not found. Please provide a valid container_name.")
+
+
+def _validate_item(item):
+    _validate_object(item, "Item not found. Please provide a valid item_name.")
+
+
+def _validate_policy(policy):
+    _validate_object(policy, "Policy not found. Please provide a valid policy_name.")
```
Comment (tjprescott) [inline src/command_modules/azure-cli-backup/azure/cli/command_modules/backup/custom.py]: It doesn't look like this addresses the key parts of #4502.

> In this case, it was not clear to me that a container and item had been created by the enable-for-vm command, so I assumed that these names were the names that would be created. 

#4502 calls for this to distinction to be made.

And then...

> The parameter should definitely accept the format of name that the container object actually uses.

I don't see that this change does that.  If the user programmatically obtains the name in the native format, the command should not choke on that.  The behavior should be similar to other commands in the CLI that accept names or IDs.  If it detects that the user supplied the long form of the name, it should parse that long form into the individual name parameters.  This scenario should also definitely be covered by a test.

File: src/command_modules/azure-cli-backup/HISTORY.rst (hunk 1 above)
Comment (dragonfly91) [reply src/command_modules/azure-cli-backup/HISTORY.rst]: @tjprescott, good catch! I'm partially fixing the bug: supplying non-existent name would not cause an issue now. However, still not handling natively formatted name. Will take it up in the next release. Will update the release history to reflect the same.

File: src/command_modules/azure-cli-backup/azure/cli/command_modules/backup/custom.py (hunk 3 above)
Comment (dragonfly91) [reply src/command_modules/azure-cli-backup/azure/cli/command_modules/backup/custom.py]: Yes, @tjprescott , as I said earlier, this change partially addresses #4502. Updated the history to reflect the correct changes. Will roll out the complete fix for #4502 in the next release.

**Comment by dragonfly91** [comment]
Hi @tjprescott , I've updated the PR description to correctly reflect the scope of the changes made by these commits. In short, this PR fully fixes #4960 and partially fixes #4502. Please let me know if there are blockers for review.

**Pull request merged**